_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+
+<!--more-->
+
+### Delivery guarantees
+
+In distributed systems like Kafka, any component can fail independently and you have to deal with this fact.
+For example, a broker may crash or a network failure may happen while the producer is sending a message to a topic.
+
+Depending on the configuration, you can have three different delivery semantics:
+
+- **at-most-once**: the producer does not retry when there is a timeout or error (data loss risk)
+- **at-least-once**: the producer retries but the broker can fail after writing but before sending the ack (duplicate messages risk)
+- **exactly-once**: even if the producer retries sending a message, it is written exactly once (idempotency)
+
+With at-least-once, there is no way for the producer to know the nature of the failure, so it simply assumes that the message was not written and retries.
+There are use cases in which lost or duplicated messages are acceptable, but in others this would cause serious business problems (e.g. a duplicate debit from a bank account).
+
+To be fair, there is no such thing as exactly-once delivery in distributed systems ([two generals' problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem)), but we can fake it by implementing idempotent log append.
+Another way to achieve EOS is to apply a deduplication logic at the consumer side, but this is a stateful logic that needs to be replicated in every consumer.
+
+> Idempotence is the property of certain operations in mathematics and computer science whereby they can be applied multiple times without changing the result beyond the initial application.
+
+Since Kafka 3.0, the producer enables the stronger delivery guarantees by default (`acks=all`, `enable.idempotence=true`).
+In addition to durability, this provides partition level message ordering and duplicates protection.
+
+When the idempotent producer is enabled, the broker registers the producer id (PID) and the epoch (producer session).
+A sequence number is assigned to each record when the batch is first added to a produce request and never changed, even if the batch is resent.
+The broker keeps track of the sequence number per producer and partition, periodically storing this information in a `.snapshot` file.
+Whenever a new batch arrives, the broker checks if the last received number is equal to the first batch number plus one, then the batch is acknowledged, otherwise it is rejected.
+
+Unfortunately, the idempotent producer does not guarantee duplicate protection across restarts and when you need to write to multiple partitions as a single unit of work.
+This is the typical scenario with many stream processing applications that run read-process-write cycles.
+In this case, you can make these cycles atomic using the transaction support of the Kafka Producer or Kafka Streams APIs.
+
+The Kafka Producer API is low level, so you need to ensure the right configuration and to commit offsets within the scope of the transaction.
```

Comment:
This is correct, but I'm not sure how many readers would really understand what it means in practice. If you don't then it's a bit frustrating because you don't elaborate on what the right configuration is, or what commit within transaction scope means. So I wonder if we can use some more general language rather than this paragraph. How about:

> The transaction support exposed by the Producer API is low level, and needs to be used with carefully in order to actually get transactional semantics. For those using Kafka Streams it is much simpler: setting `processing.guarantee=exactly_once_v2` will enable EOS on your existing topology.

Reply:
Agree and the basic example shows the correct way of using the Producer API.